core/templates/dev/head/services/contextual/UrlService.ts

```diff
@@ -17,169 +17,199 @@
-        }
-      );
-      return params;
-    },
-    isIframed: function() {
-      var pathname = this.getPathname();
-      var urlParts = pathname.split('/');
-      return urlParts[1] === 'embed';
-    },
-    getPathname: function() {
-      return this.getCurrentLocation().pathname;
-    },
-    getTopicIdFromUrl: function() {
-      var pathname = this.getPathname();
-      if (pathname.match(/\/topic_editor\/(\w|-){12}/g)) {
-        return pathname.split('/')[2];
-      }
-      throw Error('Invalid topic id url');
-    },
-    getTopicNameFromLearnerUrl: function() {
-      var pathname = this.getPathname();
-      if (pathname.match(/\/(story|topic|subtopic|practice_session)/g)) {
-        return decodeURIComponent(pathname.split('/')[2]);
-      }
-      throw Error('Invalid URL for topic');
-    },
-    getClassroomNameFromUrl: function() {
-      var pathname = this.getPathname();
-      if (pathname.match(/\/classroom/g)) {
-        return decodeURIComponent(pathname.split('/')[2]);
-      }
-      throw Error('Invalid URL for classroom');
-    },
-    getSubtopicIdFromUrl: function() {
-      var pathname = this.getPathname();
-      var argumentsArray = pathname.split('/');
-      if (pathname.match(/\/subtopic/g) && argumentsArray.length === 4) {
-        return decodeURIComponent(argumentsArray[3]);
-      }
-      throw Error('Invalid URL for subtopic');
-    },
-    getStoryIdFromUrl: function() {
-      var pathname = this.getPathname();
-      if (pathname.match(/\/(story_editor|review_test)\/(\w|-){12}/g)) {
-        return pathname.split('/')[2];
-      }
-      throw Error('Invalid story id url');
-    },
-    getStoryIdFromViewerUrl: function() {
-      var pathname = this.getPathname();
-      if (pathname.match(/\/story\/(\w|-){12}/g)) {
-        return pathname.split('/')[2];
-      }
-      throw Error('Invalid story id url');
-    },
-    getStoryIdInPlayer: function() {
-      var query = this.getCurrentQueryString();
-      var queryItems = query.split('&');
-      for (var i = 0; i < queryItems.length; i++) {
-        var part = queryItems[i];
-        if (part.match(/\?story_id=((\w|-){12})/g)) {
-          return part.split('=')[1];
-        }
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+import { WindowRef } from 'services/contextual/WindowRefService';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class UrlService {
+  constructor(private windowRef: WindowRef) {}
+
+  // This function is for testing purposes (to mock $window.location)
+  getCurrentLocation(): Location {
+    return this.windowRef.nativeWindow.location;
+  }
+
+  getCurrentQueryString(): string {
+    return this.getCurrentLocation().search;
+  }
+  /* As params[key] is overwritten, if query string has multiple fieldValues
+     for same fieldName, use getQueryFieldValuesAsList(fieldName) to get it
+     in array form. */
+  getUrlParams(): Object {
```

Comment:
I agree that we should be using ```Object``` in a more detailed way, That's a good idea but i'll really slow down our migration progress..alternatively we have a a todo for any(7165), similarly we can have a todo for ```objects/Interfaces```...becuase it'd be better to have a seperate folder and create ```Interfaces``` like ```IDictState``` for all the objects. @bansalnitish What are your thoughts on it?

Reply:
@Showtim3 Time spent to do it right the first time is, typically, time well spent -- especially if this is an issue that recurs across multiple files. If you can find a way to do it properly here, then subsequent iterations will be easy. Otherwise, you might "finish" a bit earlier, but then you would need to come back and redo all the files again to fix the TODO.

So I suggest finding a way to do this properly from the get-go. Once you do it for one example correctly, then that pattern can be replicated in other places. Does that sound good?

Reply:
@seanlip What I had in mind was that the angular migration part is tough & has more priority in the short term while adding the exact data type is comparatively easy. We have #7165 for todo's. I thought when the migration part is done, we can give some examples as how to change any to the exact data type and then open up 7165 as a beginner-friendly bug & that way we'll have plenty of that and it won't take up too much of time. But if that's isnt feasible, we can do that along the way but it'll slow us down a lot and there are around 200 services to be migrated.
//cc @bansalnitish

Reply:
Understood, thanks for explaining @Showtim3.

The point still stands, though, at least for this service. Let's at least get some idea of where the difficulties are and to what extent it will slow us down, then we have some data to make a longer-term decision (and we can talk with @bansalnitish about that). In general we should avoid creating TODOs, so the practice that you are suggesting is actually somewhat unusual.

Reply:
I understand the point being made here. I'm meeting with @bansalnitish this weekend; let's discuss it then.

Until then please avoid introducing TODOs where they can be avoided. Keeping the codebase clean is more important than short-term expediency.

Reply:
Sounds good.

Reply:
I feel I agree with @seanlip to avoid adding new TODOs but I feel that if something is too time consuming and it is complicated to add a type now, I am ok with adding a TODO in that case.